rust/src/wallet_service_client.rs

```diff
@@ -0,0 +1,400 @@
+    })?;
+
+    Ok(claims)
+}
+
+/// Internal function to create TokenInfo from JWT
+fn create_token_info(token: String) -> Result<TokenInfo, WalletServiceError> {
+    let claims = parse_jwt_claims(&token)?;
+    let expires_at = OffsetDateTime::from_unix_timestamp(claims.exp as i64).map_err(|e| {
+        WalletServiceError::JwtParseError(format!("Invalid expiration timestamp: {}", e))
+    })?;
+
+    Ok(TokenInfo {
+        token,
+        claims,
+        expires_at,
+    })
+}
+
+#[derive(uniffi::Object)]
+pub struct WalletServiceClient {
+    client: Client,
+    base_url: String,
+    token_info: Arc<Mutex<Option<TokenInfo>>>,
+}
+
+#[uniffi::export]
+impl WalletServiceClient {
+    #[uniffi::constructor]
+    pub fn new(base_url: String) -> Self {
+        Self {
+            client: Client::new(),
+            base_url,
+            token_info: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Returns the current client ID (sub claim from JWT)
+    pub fn get_client_id(&self) -> Option<String> {
+        if let Ok(guard) = self.token_info.lock() {
+            guard.as_ref().map(|info| info.claims.sub.clone())
+        } else {
+            None
+        }
+    }
+
+    /// Returns true if the current token is valid and not expired
+    pub fn is_token_valid(&self) -> bool {
+        if let Ok(guard) = self.token_info.lock() {
+            if let Some(token_info) = guard.as_ref() {
+                token_info.expires_at > OffsetDateTime::now_utc()
+            } else {
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    pub async fn login(&self, jwk: &str) -> Result<String, WalletServiceError> {
+        // Parse the JWK string into a Value to ensure it's valid JSON
+        let jwk_value: Value = serde_json::from_str(jwk)
+            .map_err(|e| WalletServiceError::InvalidJson(e.to_string()))?;
+
+        // Make POST request to /login endpoint
+        let response = self
+            .client
+            .post(format!("{}/login", self.base_url))
+            .header("Content-Type", "application/json")
+            .json(&jwk_value)
+            .send()
+            .await
+            .map_err(|e| WalletServiceError::NetworkError(e.to_string()))?;
+
+        // Check if the response was successful
+        if !response.status().is_success() {
+            let status = response.status().as_u16();
+            let error_text = response.text().await.unwrap_or_default();
+            return Err(WalletServiceError::ServerError {
+                status,
+                error_message: error_text,
+            });
+        }
+
+        // Get the response body as string
+        let token = response
+            .text()
+            .await
+            .map_err(|e| WalletServiceError::ResponseError(e.to_string()))?;
+
+        // Parse and validate the JWT
+        let token_info = create_token_info(token.clone())?;
```

Comment:
You can shortcut a lot of your manual JWT manipulation using ssi-jws and ssi-jwt:
```suggestion
        // Get the response body as bytes
        let jws_bytes: Vec<u8> = response
            .bytes()
            .await
            .map_err(|e| WalletServiceError::ResponseError(e.to_string()))?
            .into();
            
        let jws_buf = ssi::jws::JwsBuf::new(jws_bytes).map_err(|e| ...)?;
        // `to_decoded_jwt()` requires `use ssi::jwt::ToDecodedJwt;`
        let jwt_claims = jws_buf.to_decoded_jwt()
            .map_err(|e| ...)?
            .signing_bytes
            .payload;
            
        let exp: i64 = jwt_claims.get::<ssi::jwt::ExpirationTime>()
            .ok_or_else(...)?
            .0
            .as_seconds() as i64;
            
        let token_info = TokenInfo {
            token: jws_buf.into_string(),
            claims: jwt_claims,
            expires_at: todo!(),
        };

```